.github/workflows/python.yaml

```diff
@@ -31,17 +29,28 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Set up Python 3.11
-      uses: actions/setup-python@v3
+      uses: actions/setup-python@v4
       with:
         python-version: 3.11
+        cache: 'pip'
+        cache-dependency-path: |
+          dev-deps.lock
+          pyproject.toml
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
         if [ -f dev-deps.lock ]; then pip install -r dev-deps.lock; fi
         pip install -e . --no-deps
-    - name: Lint with pre-commit
-      run: |
-        pre-commit run -av
+    - name: Cache linter tool caches
+      uses: actions/cache@v3
+      with:
+        path: |
+          ~/.cache/ruff/
+          ~/.cache/mypy/
```

Comment:
I think these tools open caches in the working directory by default, so we need to set some envvars for changing the cache location - they were in the Gitlab CI, let me take a look.

Reply:
This is from the gitlab CI:

```yaml
  PRE_COMMIT_HOME: "$CI_PROJECT_DIR/.cache/pre-commit"
  MYPY_CACHE_DIR: "$CI_PROJECT_DIR/.cache/mypy"
  RUFF_CACHE_DIR: "$CI_PROJECT_DIR/.cache/ruff"
```

Since pre-commit is probably already primed on `$HOME` (I assume, from the action), substituting `~` as `$CI_PROJECT_DIR` should do the trick?